src/main/java/org/kuali/kfs/kim/document/IdentityManagementPersonDocument.java

```diff
@@ -0,0 +1,620 @@
+                            empInfo.setEmploymentRecordId(origEmpInfo.getEmploymentRecordId());
+                        }
+                    }
+                }
+                if (StringUtils.isEmpty(empInfo.getEmploymentRecordId())) {
+                    employeeRecordCounter++;
+                    empInfo.setEmploymentRecordId(employeeRecordCounter + "");
+                }
+            }
+        }
+    }
+
+    public KimTypeAttributesHelper getKimTypeAttributesHelper(String roleId) {
+        Role role = KimApiServiceLocator.getRoleService().getRole(roleId);
+        KimType kimTypeInfo = KimApiServiceLocator.getKimTypeInfoService().getKimType(role.getKimTypeId());
+        return new KimTypeAttributesHelper(kimTypeInfo);
+    }
+
+    public void setIfRolesEditable() {
+        if (CollectionUtils.isNotEmpty(getRoles())) {
+            for (PersonDocumentRole role : getRoles()) {
+                role.setEditable(validAssignRole(role));
+            }
+        }
+    }
+
+    public void setIfGroupsEditable() {
+        if (CollectionUtils.isNotEmpty(getGroups())) {
+            for (PersonDocumentGroup group : getGroups()) {
+                group.setEditable(validAssignGroup(group));
+            }
+        }
+    }
+
+    public boolean validAssignRole(PersonDocumentRole role) {
+        boolean rulePassed = true;
+        if (StringUtils.isNotEmpty(role.getNamespaceCode())) {
+            Map<String, String> additionalPermissionDetails = new HashMap<>();
+            additionalPermissionDetails.put(KimConstants.AttributeConstants.NAMESPACE_CODE, role.getNamespaceCode());
+            additionalPermissionDetails.put(KimConstants.AttributeConstants.ROLE_NAME, role.getRoleName());
+            if (!getDocumentHelperService().getDocumentAuthorizer(this).isAuthorizedByTemplate(this,
+                    KimConstants.NAMESPACE_CODE, KimConstants.PermissionTemplateNames.ASSIGN_ROLE,
+                    GlobalVariables.getUserSession().getPrincipalId(), additionalPermissionDetails, null)) {
+                rulePassed = false;
+            }
+        }
+        return rulePassed;
+    }
+
+    public boolean validAssignGroup(PersonDocumentGroup group) {
+        boolean rulePassed = true;
+        if (StringUtils.isNotEmpty(group.getNamespaceCode())) {
+            Map<String, String> additionalPermissionDetails = new HashMap<>();
+            additionalPermissionDetails.put(KimConstants.AttributeConstants.NAMESPACE_CODE, group.getNamespaceCode());
+            additionalPermissionDetails.put(KimConstants.AttributeConstants.GROUP_NAME, group.getGroupName());
+            if (!getDocumentHelperService().getDocumentAuthorizer(this).isAuthorizedByTemplate(this,
+                    KimConstants.NAMESPACE_CODE, KimConstants.PermissionTemplateNames.POPULATE_GROUP,
+                    GlobalVariables.getUserSession().getPrincipalId(), additionalPermissionDetails, null)) {
+                rulePassed = false;
+            }
+        }
```

Comment:
if the person group doesn't have a namespace code, then the group is editable on the person document?  Is that desirable?

Reply:
I was trying to set up this assign-group check to be consistent with the assign-role check in the method above it. I'm not certain why an empty-namespace check is being done here, but there should be other business rules on the Person document that prevent an empty-namespace group entry from being added (and there should not be any pre-existing empty-namespace group entries). If needed, I can update this block to return false if an empty namespace is encountered.

Reply:
sounds good to me, thanks for explaining